bndtools.m2e/src/bndtools/m2e/MavenWorkspaceRepository.java

```diff
@@ -1,256 +1,375 @@
+import org.osgi.framework.namespace.IdentityNamespace;
 import org.osgi.resource.Capability;
 import org.osgi.resource.Requirement;
-import aQute.bnd.osgi.Domain;
+import org.osgi.resource.Resource;
+import org.osgi.service.component.annotations.Component;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import aQute.bnd.osgi.Jar;
+import aQute.bnd.osgi.repository.AbstractIndexingRepository;
+import aQute.bnd.osgi.resource.CapReqBuilder;
+import aQute.bnd.osgi.resource.ResourceBuilder;
+import aQute.bnd.osgi.resource.ResourceUtils;
+import aQute.bnd.osgi.resource.ResourceUtils.ContentCapability;
+import aQute.bnd.service.RepositoryPlugin;
+import aQute.bnd.version.MavenVersion;
 import aQute.bnd.version.Version;
-import aQute.lib.collections.SortedList;
-
-public class MavenWorkspaceRepository extends AbstractMavenRepository {
+import aQute.lib.memoize.Memoize;
+import aQute.libg.glob.Glob;
+import aQute.maven.api.Revision;
+import aQute.maven.provider.POM;
+import bndtools.central.Central;
+
+@Component(service = {
+	IMavenProjectChangedListener.class, MavenWorkspaceRepository.class, RepositoryPlugin.class
+})
+public class MavenWorkspaceRepository extends
+	AbstractIndexingRepository<IProject>
+	implements IMavenProjectChangedListener, MavenRunListenerHelper, PopulatedRepository, RepositoryPlugin {
+
+	enum Kind {
+		ADDED(MavenProjectChangedEvent.KIND_ADDED),
+		CHANGED(MavenProjectChangedEvent.KIND_CHANGED),
+		REMOVED(MavenProjectChangedEvent.KIND_REMOVED);
+
+		Kind(int kind) {
+			this.kind = kind;
+		}
 
-	private final static ILogger					logger	= Logger.getLogger(MavenWorkspaceRepository.class);
+		public static Kind get(MavenProjectChangedEvent event) {
+			for (Kind value : values()) {
+				if (value.kind == event.getKind())
+					return value;
+			}
+			throw new IllegalArgumentException(String.valueOf(event.getKind()));
+		}
 
-	private boolean									inited	= false;
+		private final int kind;
+	}
 
-	private final Map<String, IMavenProjectFacade>	bsnMap	= new HashMap<>();
+	private final static Logger				logger	= LoggerFactory.getLogger(MavenWorkspaceRepository.class);
+
+	private volatile Supplier<Set<String>>								list;
+	private final Function<String, Collection<Requirement>>	bsnRequirements			= bsn -> Collections
+		.singleton(CapReqBuilder.createSimpleRequirement(IdentityNamespace.IDENTITY_NAMESPACE, bsn, null)
+			.buildSyntheticRequirement());
+	private final BiFunction<String, Version, Collection<Requirement>>	bsnAndVersionRequirements	= (
+		bsn, version) -> Collections
+			.singleton(CapReqBuilder.createSimpleRequirement(IdentityNamespace.IDENTITY_NAMESPACE, bsn, version
+				.toString())
+				.buildSyntheticRequirement());
+	private final Collection<Requirement>					identityRequirements	= Collections
+		.singleton(ResourceUtils.createWildcardRequirement());
+
+	public MavenWorkspaceRepository() {
+		super();
+		collect();
+		list = Memoize.supplier(() -> list0());
```

Comment:
I added a protected memoize method to AbstractIndexingRepository so all the memoized suppliers can be consistent. So you could use that and even override that. The default now is SoftReference memoization to allow the GC to get back our memory if really needed.

Reply:
fixed